x/superfluid/keeper/epoch_test.go

```diff
@@ -0,0 +1,100 @@
+			expectedMultiplier: sdk.MustNewDecFromStr("0.01"),
+		},
+		{
+			name:             "update LP token Osmo equivalent with pool unexpectedly deleted",
+			asset:            types.SuperfluidAsset{Denom: "gamm/pool/1", AssetType: types.SuperfluidAssetTypeLPShare},
+			poolDoesNotExist: true,
+			expectedError:    gammtypes.PoolDoesNotExistError{PoolId: 1},
+		},
+		{
+			name:               "update LP token Osmo equivalent with pool unexpectedly removed Osmo",
+			asset:              types.SuperfluidAsset{Denom: "gamm/pool/1", AssetType: types.SuperfluidAssetTypeLPShare},
+			removeStakingAsset: true,
+			expectedError:      errors.New("pool 1 has zero OSMO amount"),
+		},
+		{
+			name:               "update concentrated share Osmo equivalent successfully",
+			asset:              types.SuperfluidAsset{Denom: "cl/pool/1/1", AssetType: types.SuperfluidAssetTypeConcentratedShare},
+			expectedMultiplier: sdk.MustNewDecFromStr("1"),
+		},
+		{
+			name:             "update concentrated share Osmo equivalent with pool unexpectedly deleted",
+			asset:            types.SuperfluidAsset{Denom: "cl/pool/1/1", AssetType: types.SuperfluidAssetTypeConcentratedShare},
+			poolDoesNotExist: true,
+			expectedError:    cltypes.PoolNotFoundError{PoolId: 1},
+		},
+		{
+			name:               "update concentrated share Osmo equivalent with pool unexpectedly removed Osmo",
+			asset:              types.SuperfluidAsset{Denom: "cl/pool/1/1", AssetType: types.SuperfluidAssetTypeConcentratedShare},
+			removeStakingAsset: true,
+			expectedError:      errors.New("pool has unexpectedly removed OSMO as one of its underlying assets"),
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			ctx := suite.Ctx
+			superfluidKeeper := suite.App.SuperfluidKeeper
+
+			// Switch the default staking denom to something else if the test case requires it
+			stakeDenom := suite.App.StakingKeeper.BondDenom(ctx)
+			if tc.removeStakingAsset {
+				stakeDenom = "bar"
+			}
+			poolCoins := sdk.NewCoins(sdk.NewCoin(stakeDenom, sdk.NewInt(1000000000000000000)), sdk.NewCoin("foo", sdk.NewInt(1000000000000000000)))
+
+			// Ensure that the multiplier is zero before the test
+			multiplier := superfluidKeeper.GetOsmoEquivalentMultiplier(ctx, tc.asset.Denom)
+			suite.Require().Equal(multiplier, sdk.ZeroDec())
+
+			// Create the respective pool if the test case requires it
+			if !tc.poolDoesNotExist {
+				if tc.asset.AssetType == types.SuperfluidAssetTypeLPShare {
+					suite.PrepareBalancerPoolWithCoins(poolCoins...)
+				} else if tc.asset.AssetType == types.SuperfluidAssetTypeConcentratedShare {
+					suite.PrepareConcentratedPoolWithCoinsAndLockedFullRangePosition(stakeDenom, "foo")
+				}
+			}
+
+			// System under test
+			err := superfluidKeeper.UpdateOsmoEquivalentMultipliers(ctx, tc.asset, 1)
+
+			if tc.expectedError != nil {
+				suite.Require().Error(err)
+				suite.Require().ErrorContains(err, tc.expectedError.Error())
```

Comment:
How do we validate the outcome of `BeginUnwindSuperfluidAsset`?

Reply:
Added explicit check of multiplier and deletion of superfluid asset (the two things unwind calls) here

https://github.com/osmosis-labs/osmosis/pull/4884/commits/e183418f96267e446e5b74eb27f1193f5ce5a0c2